include/drivers/eeprom.h

```diff
@@ -0,0 +1,142 @@
+
+/**
+ * @brief EEPROM Interface
+ * @defgroup eeprom_interface EEPROM Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*eeprom_api_read)(struct device *dev, off_t offset, void *data,
+			       size_t len);
+typedef int (*eeprom_api_write)(struct device *dev, off_t offset,
+				const void *data, size_t len);
+typedef int (*eeprom_api_write_protection)(struct device *dev, bool enable);
+
+struct eeprom_driver_api {
+	eeprom_api_read read;
+	eeprom_api_write write;
+	eeprom_api_write_protection write_protection;
+	const size_t size;
+	const bool read_only;
+};
+
+/**
+ *  @brief Read data from EEPROM
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to read from.
+ *  @param data Buffer to store read data.
+ *  @param len Number of bytes to read.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_read(struct device *dev, off_t offset, void *data,
+			  size_t len);
+
+static inline int z_impl_eeprom_read(struct device *dev, off_t offset,
+				     void *data, size_t len)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->read(dev, offset, data, len);
+}
+
+/**
+ *  @brief Write data to EEPROM
+ *
+ *  Prior to the invocation of this API, the eeprom_write_protection_set needs
+ *  to be called first to disable the write protection.
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to write data to.
+ *  @param data Buffer with data to write.
+ *  @param len Number of bytes to write.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write(struct device *dev, off_t offset, const void *data,
+			   size_t len);
+
+static inline int z_impl_eeprom_write(struct device *dev, off_t offset,
+				      const void *data, size_t len)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	if (api->read_only) {
+		return -EACCES;
+	}
+
+	return api->write(dev, offset, data, len);
+}
+
+/**
+ *  @brief Enable or disable write protection for an EEPROM
+ *
+ *  @param dev EEPROM device.
+ *  @param enable enable or disable EEPROM write protection
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write_protection_set(struct device *dev, bool enable);
+
+static inline int z_impl_eeprom_write_protection_set(struct device *dev,
+						     bool enable)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->write_protection(dev, enable);
+}
+
+/**
+ *  @brief Get the size of the EEPROM in bytes
+ *
+ *  @param dev EEPROM device.
+ *
+ *  @return EEPROM size in bytes.
+ */
+__syscall size_t eeprom_get_size(struct device *dev);
+
+static inline size_t z_impl_eeprom_get_size(struct device *dev)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->size;
```

Comment:
api->size() ?

Reply:
const size_t size;

Reply:
Sorry, I overlooked that. I didn't expect something like a constant (not a function ptr) in a struct called API. But, the flash API seems to do this in the same way.